File: logic/utils/stringutils.py
import re, difflib
from collections import defaultdict
import numpy as np
from functools import lru_cache

def arr_from_sep_string(string: str, sep=','):
    return [x.strip() for x in string.split(sep)]

def arr_from_numbered_list(string: str):
    withNumbers = arr_from_sep_string(string, "\n")
    return [x.split(' ')[1] for x in withNumbers]

def extract_markdown_blocks(text):
    """
    Extract markdown blocks from a string.

    Parameters:
    text (str): The input string containing markdown blocks.

    Returns:
    list: A list of strings, each representing a markdown block.
    """
    pattern = re.compile(r"```(.*?)\n(.*?)```", re.DOTALL)
    blocks = pattern.findall(text)
    blocks = [block[1].strip() for block in blocks]
    return blocks

def markdown_to_dict(markdown: str) -> dict:
    # Regular expression to match headers
    header_regex = re.compile(r'^(#+)\s*(.*)', re.MULTILINE)
    
    # Dictionary to store the result
    result = defaultdict(str)
    
    # Find all headers
    headers = [(m.group(1).count('#'), m.group(2), m.start()) for m in header_regex.finditer(markdown)]
    
    # Sort headers by their position in the text
    headers.sort(key=lambda x: x[2])
    
    for i in range(len(headers)):
        current_header = headers[i]
        header_level, header_text, header_start = current_header
        
        # Find the end of the current header's content
        if i + 1 < len(headers):
            next_header_start = headers[i + 1][2]
            content = markdown[header_start + len(current_header[1]) + header_level + 1:next_header_start].strip()
        else:
            content = markdown[header_start + len(current_header[1]) + header_level + 1:].strip()
        
        # Add to result dictionary
        result[header_text.lower()] = content
    
    return dict(result)

def wordwise_tokenize(text):
    """Tokenizes a sequence into words, punctuation, and whitespace

    Args:
        text: The input text string.

    Returns:
        A list of tokens.
    """

    # Split the text into words and whitespace
    tokens = re.findall(r'([a-zA-Z]+|\s+|[^a-zA-Z\s]+)', text)

    return tokens

def linewise_tokenize(text: str):
    """Tokenizes a sequence into lines and \n.

    Args:
        text: The input text string.

    Returns:
        A list of tokens.
    """
    # Split the text into lines and newlines
    return text.splitlines(keepends=True)

@lru_cache(maxsize=None)
def _edit_distance(str1, str2):
    m, n = len(str1), len(str2)
    dp = [[0] * (n + 1) for _ in range(m + 1)]

    for i in range(m + 1):
        for j in range(n + 1):
            if i == 0:
                dp[i][j] = j
            elif j == 0:
                dp[i][j] = i
            elif str1[i - 1] == str2[j - 1]:
                dp[i][j] = dp[i - 1][j - 1]
            else:
                dp[i][j] = 1 + min(
                    dp[i][j - 1],      # Insert
                    dp[i - 1][j],      # Remove
                    dp[i - 1][j - 1]   # Replace
                )
    return dp[m][n]


edit_distance_cache = {}
def edit_distance(list1, list2):
    global edit_distance_cache
    if (tuple(list1), tuple(list2)) in edit_distance_cache:
        return edit_distance_cache[(tuple(list1), tuple(list2))]
    m, n = len(list1), len(list2)
    dp = [[0] * (n + 1) for _ in range(m + 1)]
    
    for i in range(m + 1):
        for j in range(n + 1):
            if i == 0:
                dp[i][j] = sum(_edit_distance("", list2[k]) for k in range(j))
            elif j == 0:
                dp[i][j] = sum(_edit_distance(list1[k], "") for k in range(i))
            else:
                cost_replace = _edit_distance(list1[i - 1], list2[j - 1])
                dp[i][j] = min(dp[i - 1][j] + _edit_distance(list1[i - 1], ""),  # Remove
                               dp[i][j - 1] + _edit_distance("", list2[j - 1]),  # Insert
                               dp[i - 1][j - 1] + cost_replace)                # Replace
    edit_distance_cache[(tuple(list1), tuple(list2))] = dp[m][n]
    return dp[m][n]

def smith_waterman_tokens(key, query, match_score=3, mismatch_penalty=-3, gap_penalty=-2, processing=None, diagonal=None, key_token_depth=None, query_token_depth=None):
    # Initialize the scoring matrix and the traceback matrix
    rows = len(key) + 1
    cols = len(query) + 1
    scoring_matrix = np.zeros((rows, cols), dtype=int)
    traceback_matrix = np.zeros((rows, cols), dtype=int)

    max_score = 0
    max_pos = (0, 0)

    # Fill the scoring matrix and the traceback matrix
    for i in range(1, rows):
        for j in range(1, cols):
            t1, t2 = key[i-1], query[j-1]
            if processing:
                t1 = processing(t1)
                t2 = processing(t2)
            multiplier = 1 + len(max(t1, t2, key=len))
            match = scoring_matrix[i-1, j-1] + (match_score * multiplier if t1 == t2 else mismatch_penalty * multiplier)
            delete = scoring_matrix[i-1, j] + gap_penalty * multiplier
            insert = scoring_matrix[i, j-1] + gap_penalty * multiplier
            scoring_matrix[i, j] = max(0, match, delete, insert)

            if scoring_matrix[i, j] == match:
                traceback_matrix[i, j] = 1  # Diagonal
            elif scoring_matrix[i, j] == delete:
                traceback_matrix[i, j] = 2  # Up
            elif scoring_matrix[i, j] == insert:
                traceback_matrix[i, j] = 3  # Left

            if scoring_matrix[i, j] >= max_score:
                max_score = scoring_matrix[i, j]
                max_pos = (i, j)

    # Traceback to get the optimal alignment
    align1, align2 = [], []
    i, j = max_pos
    while scoring_matrix[i, j] != 0:
        if traceback_matrix[i, j] == 1:
            align1.insert(0, key[i-1])
            align2.insert(0, query[j-1])
            i -= 1
            j -= 1
        elif traceback_matrix[i, j] == 2:
            align1.insert(0, key[i-1])
            align2.insert(0, '-')
            i -= 1
        elif traceback_matrix[i, j] == 3:
            align2.insert(0, query[j-1])
            align1.insert(0, '-')
            j -= 1

    return align1, align2, max_score, scoring_matrix

def find_most_similar_substring_naive(corpus, query, match_score=3, mismatch_penalty=-3, gap_penalty=-1):
    corpus_tokens = linewise_tokenize(corpus)
    query_tokens = linewise_tokenize(query)
    corpus_align, query_align, max_score, _= smith_waterman_tokens(corpus_tokens, query_tokens, match_score, mismatch_penalty, gap_penalty, processing=lambda x: x.lower().strip())
    return ''.join(corpus_align), max_score

def score(a, b):
    max_len = max(len(''.join(a)), len(''.join(b)))
    if max_len == 0: return 1
    diff = edit_distance(a, b)
    return 1 - (diff / max_len)

def weighted_score(a, b):
    return score(a, b) * np.mean([len(a), len(b)])

db_kmers_cache = {}

def ktuple_matching(query, database, k, part_processing=None):
    global db_kmers_cache
    if not part_processing:
        part_processing = lambda x: x
    if (''.join(database), k) in db_kmers_cache:
        database_kmers = db_kmers_cache[(''.join(database), k)]
    else:
        database_kmers = defaultdict(list)
        for i in range(len(database) - k + 1):
            kmer = ''.join(part_processing(part) for part in database[i:i+k])
            if ''.join(kmer) ==  '': continue
            database_kmers[kmer].append(i)
        db_kmers_cache[(''.join(database), k)] = database_kmers
    
    matches = []
    for i in range(len(query) - k + 1):
        kmer = ''.join(part_processing(part) for part in query[i:i+k])
        if ''.join(kmer) == '': continue
        if kmer in database_kmers:
            for db_pos in database_kmers[kmer]:
                matches.append((i, db_pos))
    return matches

def find_top_diagonal(matches, query, database, k):
    diagonal_scores = defaultdict(int)
    for q_pos, db_pos in matches:
        diag = q_pos - db_pos
        diagonal_scores[diag] += np.mean([weighted_score(q, db) for q, db in zip(query[q_pos:q_pos+k], database[db_pos:db_pos+k])])
    top_diagonal = max(diagonal_scores.items(), key=lambda item: item[1], default=(None, 0))
    return top_diagonal[0]

def smith_waterman_diagonal(
        query,
        database,
        match_score=3,
        mismatch_penalty=3,
        gap_penalty=2,
        processing=None,
        tolerance=0.75,
        diag=0,
        band_width=5,
        db_token_depth_roc=None,
        query_token_depth_roc=None
    ):
    m, n = len(query), len(database)
    
    scoring_matrix, traceback_matrix = defaultdict(int), defaultdict(int)
    if diag >= 0:
        start_row = 1
        start_col = -diag
    else:
        start_row = diag
        start_col = 1

    max_score = 0
    max_pos = (0, 0)
    covered = []
    while start_row <= m and start_col <= n:
        for offset in range(-band_width, band_width + 1):
            row, col = start_row, start_col + offset
            if not (0 < row <= m and 0 < col <= n): continue
            q, d = query[row-1], database[col-1]
            covered.append((row - 1, col - 1))
            if processing:
                q, d = processing(q), processing(d)
            multiplier = 1 + np.mean([len(q), len(d)]) / 2
            match = scoring_matrix[(row-1, col-1)] + (match_score if score(q, d) >= tolerance else -mismatch_penalty) * multiplier
            insert = scoring_matrix[(row, col-1)] - gap_penalty * multiplier
            scoring_matrix[(row, col)] = max(match, insert)

            if scoring_matrix[(row, col)] == match:
                traceback_matrix[(row, col)] = 1
            elif scoring_matrix[(row, col)] == insert:
                traceback_matrix[(row, col)] = 3
            
            if scoring_matrix[(row, col)] >= max_score:
                max_score = scoring_matrix[(row, col)]
                max_pos = (row, col)
        start_row += 1
        start_col += 1

    # Traceback to get the optimal alignment
    align1, align2 = [], []
    i, j = max_pos
    while scoring_matrix[(i, j)] > 0:
        q, d = query[i - 1], database[j - 1]
        if traceback_matrix[(i, j)] == 1:
            align1.insert(0, q)
            align2.insert(0, d)
            i -= 1
            j -= 1
        elif traceback_matrix[(i, j)] == 3:
            align2.insert(0, d)
            align1.insert(0, '-')
            j -= 1
    return align2, max_score, covered

def fasta_algorithm(database, query, k=4, n=3, band_width=5, match_score=3, mismatch_penalty=3, gap_penalty=1, match_processing=None, dp_processing=None, db_token_depth_roc=None, query_token_depth_roc=None):
    matches = ktuple_matching(query, database, k, match_processing)
    top_diagonal = find_top_diagonal(matches, query, database, k)
    if top_diagonal is None: return None, float('-inf')
    best_alignment, best_score, covered = smith_waterman_diagonal(
        query,
        database,
        match_score=match_score,
        mismatch_penalty=mismatch_penalty,
        gap_penalty=gap_penalty,
        processing=dp_processing,
        diag=top_diagonal,
        band_width=band_width,
        db_token_depth_roc=db_token_depth_roc,
        query_token_depth_roc=query_token_depth_roc
    )
    if best_alignment is None:
        return None, float('-inf')
    return best_alignment, best_score

def fuzzy_find(query, database):
    # 1. First pass with linewise tokenization
    linewise_tokenized_database = linewise_tokenize(database)
    linewise_tokenized_query = linewise_tokenize(query)

    result, score = fasta_algorithm(
        linewise_tokenized_database,
        linewise_tokenized_query,
        k=1,
        n=1,
        band_width=10,
        match_score=3,
        mismatch_penalty=3,
        gap_penalty=0.1,
        match_processing=lambda x: x.strip().replace(' ', ''),
        dp_processing=lambda x: wordwise_tokenize(x.strip().replace(' ', '')),
    )

    # 2. Second pass with wordwise tokenization
    worwise_tokenized_database = wordwise_tokenize(''.join(result)) if result is not None else wordwise_tokenize(database)
    wordwise_tokenized_query = wordwise_tokenize(query)

    result, score = fasta_algorithm(
        worwise_tokenized_database,
        wordwise_tokenized_query,
        k=max(1, min(3, len(wordwise_tokenized_query) - 1, len(worwise_tokenized_database) - 1)),
        n=1,
        band_width=18,
        match_score=5,
        mismatch_penalty=3,
        gap_penalty=0.3,
        match_processing=lambda x: x.strip(),
        dp_processing=lambda x: x.strip(),
    )

    if result is None:
        return None, float('-inf')
    return ''.join(result), score

def get_diffs(a, b):
    diffs = [diff for diff in difflib.ndiff(a.splitlines(), b.splitlines())]
    diffs_groups = []
    current_group = []
    for diff in diffs:
        if diff.startswith(' '):
            if current_group:
                diffs_groups.append(current_group)
                current_group = []
        elif diff.startswith('-') or diff.startswith('+'):
            current_group.append(diff)
    diffs = []
    for group in diffs_groups:
        # Turn group into an array of tuples (find, replace)
        find = '\n'.join([line[1:] for line in group if line.startswith('-')])
        replace = '\n'.join([line[1:] for line in group if line.startswith('+')])
        diffs.append((find, replace))
    return diffs

def parse_diff(diff_string):
    lines = diff_string.splitlines()
    groups = []
    current_group = []
    in_edit = False

    for line in lines:
        if line.startswith('+') or line.startswith('-'):
            if not in_edit and current_group:
                groups.append(current_group)
                current_group = []
            current_group.append(line)
            in_edit = True
        else:
            if in_edit and current_group:
                groups.append(current_group)
                current_group = []
            current_group.append(line)
            in_edit = False
    
    if current_group:
        groups.append(current_group)
    
    merged_groups = []
    for group in groups:
        if not (group[0].startswith('+') or group[0].startswith('-')):
            merged_groups.append('\n'.join(group))
        else:
            delete, insert = '', ''
            for line in group:
                if line.startswith('-'):
                    delete += line[1:] + '\n'
                elif line.startswith('+'):
                    insert += line[1:] + '\n'
            merged_groups.append((delete, insert))

    return merged_groups

def skwonk(database: str, diff: str):
    hunks = re.split(r"@@.*?@@", diff)
    i = 0
    for hunk in hunks:
        original = ''.join([line for line in hunk.splitlines(keepends=True) if not line.startswith('+')])
        # 1. Zone in on the string we are going to be editing
        zone, _ = fuzzy_find(original, database[i:])
        i = database.find(zone) + len(zone)
        original_zone = zone
        if zone is None:
            return database
        
        # 2. Parse the diff into groups
        diff_groups = parse_diff(hunk)
        # 3. Find and replace
        for i in range(len(diff_groups)):
            group = diff_groups[i]
            if not isinstance(group, tuple): continue
            find, replace = group
            if find == '':
                # Insertion
                prior_context = fuzzy_find(diff_groups[i - 1], zone)[0] if i > 0 else None
                further_context = fuzzy_find(diff_groups[i + 1], zone)[0] if i < len(diff_groups) - 1 else None
                insertion_index = zone.find(prior_context) + len(prior_context) if prior_context else None
                if not insertion_index:
                    insertion_index = zone.find(further_context) if further_context else 0
                zone = zone[:insertion_index] + '\n' + replace + zone[insertion_index:]
            else:
                # Fuzzy find and replace
                fuzz, _ = fuzzy_find(find, zone)
                if fuzz is None: continue
                zone = zone.replace(fuzz, replace, 1)
        database = database.replace(original_zone, zone, 1)
    return database

def find_most_similar_file_name(files, query):
    return max(
        (file for file in files),
        key=lambda file: fasta_algorithm(file, query, k=6)[1]
    )

def compute_nested_levels(code_lines: list[str], indent_is_relevant: bool = True) -> list[int]:
    nested_levels = []
    brace_level = 0
    paren_level = 0
    indent_level = 0
    current_level = 0
    prev_indent_spaces = 0

    for line in code_lines:
        if line.strip() == '':
            nested_levels.append(current_level)
            continue
        stripped_line = line.rstrip()
        # Update the brace and parenthesis levels for the next line
        brace_level += stripped_line.count('{')
        brace_level -= stripped_line.count('}')
        paren_level += stripped_line.count('(')
        paren_level -= stripped_line.count(')')
        if indent_is_relevant:
            leading_spaces = len(line) - len(line.lstrip())
            indent_level += np.sign(leading_spaces - prev_indent_spaces)
            prev_indent_spaces = leading_spaces
        # Make sure levels do not go negative
        if brace_level < 0:
            brace_level = 0
        if paren_level < 0:
            paren_level = 0
        
        # Calculate the current line's nested level
        current_level = brace_level + paren_level + indent_level
        nested_levels.append(current_level)
    return nested_levels

File: newlogic/utils/stringutils.js
const _ = require('lodash');
const re = require('re');
const { lruCache } = require('lodash');
const { default: difflib } = require('re');

const arrFromSepString = (string, sep = ',') => string.split(sep).map(x => x.trim());

const arrFromNumberedList = (string) => arrFromSepString(string, "\n").map(x => x.split(' ')[1]);

const extractMarkdownBlocks = (text) => {
    const pattern = /```(.*?)\n(.*?)```/gs;
    const blocks = Array.from(text.matchAll(pattern)).map(match => match[2].trim());
    return blocks;
};

const markdownToDict = (markdown) => {
    const headerRegex = /^#+\s*(.*)/gm;
    const result = {};
    let match;
    const headers = [];

    while ((match = headerRegex.exec(markdown)) !== null) {
        headers.push([match[0].length, match[1], match.index]);
    }

    headers.sort((a, b) => a[2] - b[2]);

    for (let i = 0; i < headers.length; i++) {
        const [headerLevel, headerText, headerStart] = headers[i];
        const nextHeaderStart = i + 1 < headers.length ? headers[i + 1][2] : markdown.length;
        const content = markdown.slice(headerStart + headerText.length + headerLevel, nextHeaderStart).trim();
        result[headerText.toLowerCase()] = content;
    }

    return result;
};

const wordwiseTokenize = (text) => text.match(/[a-zA-Z]+|\s+|[^a-zA-Z\s]+/g) || [];

const linewiseTokenize = (text) => text.split(/\r?\n/);

const _editDistance = _.memoize((str1, str2) => {
    const m = str1.length, n = str2.length;
    const dp = Array.from({ length: m + 1 }, () => Array(n + 1).fill(0));

    for (let i = 0; i <= m; i++) {
        for (let j = 0; j <= n; j++) {
            if (i === 0) dp[i][j] = j;
            else if (j === 0) dp[i][j] = i;
            else dp[i][j] = str1[i - 1] === str2[j - 1] ? dp[i - 1][j - 1] : 1 + Math.min(dp[i - 1][j], dp[i][j - 1], dp[i - 1][j - 1]);
        }
    }
    return dp[m][n];
});

const editDistanceCache = new Map();
const editDistance = (list1, list2) => {
    const key = [list1, list2].toString();
    if (editDistanceCache.has(key)) return editDistanceCache.get(key);
    const m = list1.length, n = list2.length;
    const dp = Array.from({ length: m + 1 }, () => Array(n + 1).fill(0));

    for (let i = 0; i <= m; i++) {
        for (let j = 0; j <= n; j++) {
            if (i === 0) dp[i][j] = _.sum(_.range(j).map(k => _editDistance('', list2[k])));
            else if (j === 0) dp[i][j] = _.sum(_.range(i).map(k => _editDistance(list1[k], '')));
            else dp[i][j] = Math.min(dp[i - 1][j] + _editDistance(list1[i - 1], ''), dp[i][j - 1] + _editDistance('', list2[j - 1]), dp[i - 1][j - 1] + _editDistance(list1[i - 1], list2[j - 1]));
        }
    }
    editDistanceCache.set(key, dp[m][n]);
    return dp[m][n];
};

// Function to calculate the score based on edit distance
const score = (a, b) => {
    const maxLen = Math.max(a.join('').length, b.join('').length);
    if (maxLen === 0) return 1;
    const diff = editDistance(a, b);
    return 1 - diff / maxLen;
};

const weightedScore = (a, b) => score(a, b) * _.mean([a.length, b.length]);


let dbKmersCache = {};

function ktupleMatching(query, database, k, partProcessing = (x) => x) {
    let databaseKmers;

    const cacheKey = [database.join(''), k].toString();
    if (cacheKey in dbKmersCache) {
        databaseKmers = dbKmersCache[cacheKey];
    } else {
        databaseKmers = {};
        for (let i = 0; i <= database.length - k; i++) {
            let kmer = database.slice(i, i + k).map(partProcessing).join('');
            if (kmer === '') continue;
            if (!databaseKmers[kmer]) {
                databaseKmers[kmer] = [];
            }
            databaseKmers[kmer].push(i);
        }
        dbKmersCache[cacheKey] = databaseKmers;
    }

    let matches = [];
    for (let i = 0; i <= query.length - k; i++) {
        let kmer = query.slice(i, i + k).map(partProcessing).join('');
        if (kmer === '') continue;
        if (kmer in databaseKmers) {
            for (let dbPos of databaseKmers[kmer]) {
                matches.push([i, dbPos]);
            }
        }
    }
    return matches;
}

function findTopDiagonal(matches, query, database, k) {
    let diagonalScores = {};
    for (let [qPos, dbPos] of matches) {
        let diag = qPos - dbPos;
        if (!(diag in diagonalScores)) {
            diagonalScores[diag] = 0;
        }
        diagonalScores[diag] += query.slice(qPos, qPos + k)
            .map((q, i) => weightedScore(q, database[dbPos + i]))
            .reduce((a, b) => a + b, 0) / k;
    }
    let topDiagonal = Object.entries(diagonalScores).reduce(
        (maxEntry, entry) => entry[1] > maxEntry[1] ? entry : maxEntry, [null, 0]
    );
    return topDiagonal[0];
}

function smithWatermanDiagonal({
    query, database, matchScore = 3, mismatchPenalty = 3, gapPenalty = 2, 
    processing = null, tolerance = 0.75, diag = 0, bandWidth = 5, 
    dbTokenDepthRoc = null, queryTokenDepthRoc = null
}) {
    const m = query.length, n = database.length;
    let scoringMatrix = {}, tracebackMatrix = {};
    let startRow = diag >= 0 ? 1 : diag;
    let startCol = diag >= 0 ? -diag : 1;

    let maxScore = 0;
    let maxPos = [0, 0];
    let covered = [];

    while (startRow <= m && startCol <= n) {
        for (let offset = -bandWidth; offset <= bandWidth; offset++) {
            let row = startRow, col = startCol + offset;
            if (!(0 < row && row <= m && 0 < col && col <= n)) continue;

            let q = query[row - 1], d = database[col - 1];
            covered.push([row - 1, col - 1]);

            if (processing) {
                q = processing(q);
                d = processing(d);
            }
            let multiplier = 1 + (q.length + d.length) / 2;
            let match = (scoringMatrix[[row - 1, col - 1]] || 0) + 
                        (score(q, d) >= tolerance ? matchScore : -mismatchPenalty) * multiplier;
            let insert = (scoringMatrix[[row, col - 1]] || 0) - gapPenalty * multiplier;

            scoringMatrix[[row, col]] = Math.max(match, insert);

            if (scoringMatrix[[row, col]] === match) {
                tracebackMatrix[[row, col]] = 1;
            } else if (scoringMatrix[[row, col]] === insert) {
                tracebackMatrix[[row, col]] = 3;
            }

            if (scoringMatrix[[row, col]] >= maxScore) {
                maxScore = scoringMatrix[[row, col]];
                maxPos = [row, col];
            }
        }
        startRow++;
        startCol++;
    }

    let align1 = [], align2 = [];
    let [i, j] = maxPos;
    while ((scoringMatrix[[i, j]] || 0) > 0) {
        let q = query[i - 1], d = database[j - 1];
        if (tracebackMatrix[[i, j]] === 1) {
            align1.unshift(q);
            align2.unshift(d);
            i--; j--;
        } else if (tracebackMatrix[[i, j]] === 3) {
            align2.unshift(d);
            align1.unshift('-');
            j--;
        }
    }
    return [align2, maxScore, covered];
}

function fastaAlgorithm({
    database, query, k = 4, n = 3, bandWidth = 5, matchScore = 3, mismatchPenalty = 3, 
    gapPenalty = 1, matchProcessing = null, dpProcessing = null, 
    dbTokenDepthRoc = null, queryTokenDepthRoc = null
}) {
    const matches = ktupleMatching(query, database, k, matchProcessing);
    const topDiagonal = findTopDiagonal(matches, query, database, k);
    if (topDiagonal === null) return [null, -Infinity];

    const [bestAlignment, bestScore, covered] = smithWatermanDiagonal({
        query, database, matchScore, mismatchPenalty, gapPenalty, 
        processing: dpProcessing, diag: topDiagonal, bandWidth, 
        dbTokenDepthRoc, queryTokenDepthRoc
    });

    if (!bestAlignment) return [null, -Infinity];
    return [bestAlignment, bestScore];
}

function fuzzyFind(query, database) {
    const linewiseTokenizedDatabase = linewiseTokenize(database);
    const linewiseTokenizedQuery = linewiseTokenize(query);

    let [result, score] = fastaAlgorithm({
        database: linewiseTokenizedDatabase,
        query: linewiseTokenizedQuery,
        k: 1,
        n: 1,
        bandWidth: 10,
        matchScore: 3,
        mismatchPenalty: 3,
        gapPenalty: 0.1,
        matchProcessing: (x) => x.trim().replace(/ /g, ''),
        dpProcessing: (x) => wordwiseTokenize(x.trim().replace(/ /g, ''))
    });

    const wordwiseTokenizedDatabase = result ? wordwiseTokenize(result.join('')) : wordwiseTokenize(database);
    const wordwiseTokenizedQuery = wordwiseTokenize(query);

    [result, score] = fastaAlgorithm({
        database: wordwiseTokenizedDatabase,
        query: wordwiseTokenizedQuery,
        k: Math.max(1, Math.min(3, wordwiseTokenizedQuery.length - 1, wordwiseTokenizedDatabase.length - 1)),
        n: 1,
        bandWidth: 18,
        matchScore: 5,
        mismatchPenalty: 3,
        gapPenalty: 0.3,
        matchProcessing: (x) => x.trim(),
        dpProcessing: (x) => x.trim()
    });

    if (!result) return [null, -Infinity];
    return [result.join(''), score];
}

function getDiffs(a, b) {
    const diffs = Array.from(Diff.diffLines(a, b));
    let diffsGroups = [];
    let currentGroup = [];
    diffs.forEach(diff => {
        if (!diff.added && !diff.removed) {
            if (currentGroup.length > 0) {
                diffsGroups.push(currentGroup);
                currentGroup = [];
            }
        } else {
            currentGroup.push(diff);
        }
    });
    if (currentGroup.length > 0) {
        diffsGroups.push(currentGroup);
    }

    return diffsGroups.map(group => {
        const find = group.filter(line => line.removed).map(line => line.value).join('\n');
        const replace = group.filter(line => line.added).map(line => line.value).join('\n');
        return [find, replace];
    });
}

function parseDiff(diffString) {
    const lines = diffString.split('\n');
    let groups = [];
    let currentGroup = [];
    let inEdit = false;

    lines.forEach(line => {
        if (line.startsWith('+') || line.startsWith('-')) {
            if (!inEdit && currentGroup.length > 0) {
                groups.push(currentGroup);
                currentGroup = [];
            }
            currentGroup.push(line);
            inEdit = true;
        } else {
            if (inEdit && currentGroup.length > 0) {
                groups.push(currentGroup);
                currentGroup = [];
            }
            currentGroup.push(line);
            inEdit = false;
        }
    });
    if (currentGroup.length > 0) {
        groups.push(currentGroup);
    }

    return groups.map(group => {
        if (!(group[0].startsWith('+') || group[0].startsWith('-'))) {
            return group.join('\n');
        } else {
            let deleteLines = '';
            let insertLines = '';
            group.forEach(line => {
                if (line.startsWith('-')) deleteLines += line.slice(1) + '\n';
                if (line.startsWith('+')) insertLines += line.slice(1) + '\n';
            });
            return [deleteLines, insertLines];
        }
    });
}

function skwonk(database, diff) {
    const hunks = diff.split(/@@.*?@@/);
    let i = 0;

    for (let hunk of hunks) {
        const original = hunk.split('\n').filter(line => !line.startsWith('+')).join('\n');
        const [zone, ] = fuzzyFind(original, database.slice(i));
        i = database.indexOf(zone) + zone.length;
        const originalZone = zone;

        if (!zone) return database;

        const diffGroups = parseDiff(hunk);
        diffGroups.forEach((group, idx) => {
            if (!Array.isArray(group)) return;
            let [find, replace] = group;

            if (find === '') {
                const priorContext = fuzzyFind(diffGroups[idx - 1], zone)[0] || null;
                const furtherContext = fuzzyFind(diffGroups[idx + 1], zone)[0] || null;
                let insertionIndex = priorContext ? zone.indexOf(priorContext) + priorContext.length : null;
                if (insertionIndex === null) {
                    insertionIndex = furtherContext ? zone.indexOf(furtherContext) : 0;
                }
                zone = zone.slice(0, insertionIndex) + '\n' + replace + zone.slice(insertionIndex);
            } else {
                const [fuzz, ] = fuzzyFind(find, zone);
                if (fuzz !== null) {
                    zone = zone.replace(fuzz, replace);
                }
            }
        });

        database = database.replace(originalZone, zone);
    }
    return database;
}

function findMostSimilarFileName(files, query) {
    return files.reduce((bestMatch, file) => {
        const score = fastaAlgorithm({database: file, query, k: 6})[1];
        return score > bestMatch[1] ? [file, score] : bestMatch;
    }, [null, -Infinity])[0];
}

function computeNestedLevels(codeLines, indentIsRelevant = true) {
    let nestedLevels = [];
    let braceLevel = 0, parenLevel = 0, indentLevel = 0;
    let currentLevel = 0;
    let prevIndentSpaces = 0;

    codeLines.forEach(line => {
        if (line.trim() === '') {
            nestedLevels.push(currentLevel);
            return;
        }

        const strippedLine = line.trimEnd();
        braceLevel += (strippedLine.match(/{/g) || []).length;
        braceLevel -= (strippedLine.match(/}/g) || []).length;
        parenLevel += (strippedLine.match(/\(/g) || []).length;
        parenLevel -= (strippedLine.match(/\)/g) || []).length;

        if (indentIsRelevant) {
            const leadingSpaces = line.length - line.trimStart().length;
            indentLevel += Math.sign(leadingSpaces - prevIndentSpaces);
            prevIndentSpaces = leadingSpaces;
        }

        if (braceLevel < 0) braceLevel = 0;
        if (parenLevel < 0) parenLevel = 0;

        currentLevel = braceLevel + parenLevel + indentLevel;
        nestedLevels.push(currentLevel);
    });

    return nestedLevels;
}

module.exports = {
    arrFromSepString,
    arrFromNumberedList,
    extractMarkdownBlocks,
    markdownToDict,
    wordwiseTokenize,
    linewiseTokenize,
    editDistance,
    score,
    weightedScore,
    fuzzyFind,
    skwonk,
    findMostSimilarFileName,
    computeNestedLevels
};
